#NLP RAG Project - Medical Assistant
###Meskerem Goshime, January 26, 2026

## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 205.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 276.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 346.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 240.4 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.19.0 requires numpy<2.2.0,>=1.26.0, but you have numpy 2.4.1 which is incompatible.
opencv-python-headless 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.4.1 which is incompatible.
opencv-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.4.

In [ ]:
!pip install huggingface_hub==0.35.3 pandas==2.2.2 tiktoken==0.12.0 pymupdf==1.26.5 langchain==0.3.27 langchain-community==0.3.31 chromadb==1.1.1 sentence-transformers==5.1.1 numpy==2.3.3 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 6.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 564.3/564.3 kB 49.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 109.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 69.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 95.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 MB 125.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.6/486.6 kB 47.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 141.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 24.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 102.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 458.9/458.9 kB 41.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
#importing libraries
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma
from huggingface_hub import hf_hub_download
from langchain_community.llms import LlamaCpp
from langchain.chains import RetrievalQA


## Question Answering using LLM

#### Downloading and Loading the model

In [ ]:
#importing the LLM model
model_name_or_path = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
model_basename = "mistral-7b-instruct-v0.2.Q6_K.gguf"

# Download the model file from HuggingFace Hub
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

In [ ]:
#instantiate the model

llm = LlamaCpp(
    model_path=model_path,
    #temperature=0.01,
    #top_p=0.95,
    n_ctx=2300,
    n_batch=512,
    n_gpu_layers=38,
    #verbose=True,
)

AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [ ]:
#define the response function
def response(query,max_tokens=2000,temperature=0,top_p=0.95,top_k=20):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
query1 = "What is the protocol for managing sepsis in a critical care unit?"

In [ ]:
response(query1)

/tmp/ipython-input-501375750.py:2: LangChainDeprecationWarning: The method `BaseLLM.__call__` was deprecated in langchain-core 0.1.7 and will be removed in 1.0. Use :meth:`~invoke` instead.
  model_output = llm(


'\n\nSepsis is a life-threatening condition that can arise from an infection. In a critical care unit, the management of sepsis involves a systematic and rapid approach to identify and treat the underlying infection, as well as addressing the systemic inflammatory response.\n\nThe following are key steps in managing sepsis in a critical care unit:\n\n1. Early recognition and suspicion of sepsis based on clinical signs and symptoms, such as fever or hypothermia, tachycardia or bradycardia, altered mental status, and lactic acidosis.\n2. Obtain a thorough history and physical examination to identify the source of infection, such as pneumonia, urinary tract infection, or intra-abdominal infection.\n3. Initiate appropriate antibiotic therapy based on the suspected source of infection and local antibiogram data.\n4. Provide adequate fluid resuscitation to maintain adequate tissue perfusion and organ function. This may involve the use of crystalloid or colloid solutions, as well as the use o

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
query2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"


In [ ]:
response(query2)

Llama.generate: prefix-match hit


'\n\nAppendicitis is a medical condition characterized by inflammation of the appendix, a small pouch located in the lower right side of the abdomen.\n\nThe common symptoms for appendicitis include:\n\n1. Abdominal pain, usually starting around the navel and then shifting to the lower right side of the abdomen. The pain may be constant or come and go.\n2. Loss of appetite and/or weight loss.\n3. Nausea and vomiting.\n4. Fever, which may be low-grade or high.\n5. Abdominal swelling and/or tenderness.\n6. Passing gas (flatulence) or diarrhea.\n7. Painful urination or frequent urination.\n8. Difficulty moving the bowels (constipation).\n9. Pain in the back, especially on the right side.\n10. Pain in the pelvic area, especially when sitting down or walking for long periods of time.\n\nIt is important to note that not all individuals with appendicitis will experience all of these symptoms. Additionally, some of these symptoms may be present in other medical conditions, so it is important to

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
query3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"

In [ ]:
response(query3)

Llama.generate: prefix-match hit


'\n\nSudden patchy hair loss can be a distressing experience. The causes of patchy hair loss can vary from person to person. Here are some possible causes:\n\n1. Alopecia Areata: This is an autoimmune disorder that causes hair loss in round patches on the scalp, beard, or eyebrows.\n2. Traction Alopecia: This is a type of hair loss caused by pulling or traction on the hair. It can occur due to tight hairstyles like braids, ponytails, and cornrows.\n3. Nutritional Deficiencies: Hair loss can also be caused by nutritional deficiencies, particularly in iron, zinc, biotin, and vitamin D.\n4. Stress: Hair loss can also be caused by stress, which can lead to a condition called telogen effluvium.\n5. Medications: Certain medications, such as antidepressants, beta-blockers, and retinoids, can cause hair loss as a side effect.\n6. Hormonal Imbalance: Hair loss can also be caused by hormonal imbalances, particularly in thyroid hormones or androgens like testosterone.\n7. Infections: Hair loss ca

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
query4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"


In [ ]:
response(query4)

Llama.generate: prefix-match hit


"\n\nThere is no one-size-fits-all answer to this question, as the specific treatment recommendations for a person with a brain injury will depend on several factors, including:\n\n1. The severity and location of the brain injury.\n2. The age and overall health of the person with the brain injury.\n3. The presence of any other injuries or conditions that may be affecting the person's brain function.\n\nBased on these factors, a healthcare professional may recommend one or more of the following treatments for a person with a brain injury:\n\n1. Medications to manage symptoms such as pain, headache, seizures, and depression.\n2. Physical therapy to help improve strength, balance, coordination, and mobility.\n3. Occupational therapy to help improve fine motor skills, hand-eye coordination, and the ability to perform daily activities.\n4. Speech-language therapy to help improve communication skills, including speaking, listening, reading, and writing.\n5. Cognitive rehabilitation to help i

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
query5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"


In [ ]:
response(query5)

Llama.generate: prefix-match hit


'\n\nFirst and foremost, it is essential to ensure the safety and stability of the injured person. If the fracture is severe or the person is in significant pain, it may be necessary to call for emergency medical assistance.\n\nOnce the injured person has been stabilized and any immediate medical needs have been addressed, the focus can shift to their care and recovery.\n\nSome essential steps for caring for a person with a fractured leg during a hiking trip include:\n\n1. Provide adequate pain relief: Depending on the severity of the fracture, the injured person may require prescription pain medication or other forms of pain relief. It is important to follow the instructions provided by the healthcare professional regarding the administration and dosage of any prescribed pain medications.\n2. Apply ice packs to reduce swelling: Applying ice packs to the affected area can help to reduce swelling, inflammation, and pain. It is recommended to apply ice packs for 15-20 minutes at a time, 

Observations:
- The responses with LLM were decent.
- However this will not work in a medical setting as the model is not using authoritative sources.

## Question Answering using LLM with Prompt Engineering

Note:
- In the question answering with LLM section, some of the responses were very long. Therefore, I will reduce the max_tokens size to 500 from 2000 in this section.
- I will also include instruction to respond conscisely in the system prompt.
- I will also include instruction to respond as I don't have enough information when appropriate.

In [ ]:
# defining the system prompt
system_prompt = """
[INST]<<SYS>> Respond to the user question based on the user prompt.
Please be conscise when responding.
If you don't have enough information on the topic,
please respond by I don't have enough information on the topic.<</SYS>>[/INST]
"""

In [ ]:
#defining the response function
def response_prompt(query,max_tokens=500,temperature=0,top_p=0.95,top_k=20):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
#combining the system prompt with the query
prompt1 = system_prompt+"\n"+query1

In [ ]:
response_prompt(prompt1)

Llama.generate: prefix-match hit


" I don't have enough information on the specific protocol for managing sepsis in a critical care unit. However, generally speaking, early recognition and prompt initiation of appropriate antibiotics are key components of sepsis management in a critical care setting. Other important interventions may include fluid resuscitation, vasopressor support, and mechanical ventilation as needed to maintain adequate oxygenation and circulation. Additionally, ongoing assessment and monitoring of the patient's clinical status, laboratory values, and response to treatment are essential components of effective sepsis management in a critical care setting."

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
prompt2 = system_prompt+"\n"+query2

In [ ]:
response_prompt(prompt2)

Llama.generate: prefix-match hit


"\n\nI. Common Symptoms for Appendicitis:\n1. Sudden loss of appetite or feeling sick to your stomach.\n2. Pain in the lower right-hand side of your abdomen (belly). This pain may start as a mild discomfort, but it can quickly become severe and constant.\n3. Fever, which is often accompanied by chills.\n4. Nausea and vomiting.\n5. Diarrhea or constipation.\n6. Inability to pass gas (flatulence).\n7. Painful walking or standing for long periods of time.\n8. Swelling or tenderness in the lower right-hand side of your abdomen.\n9. Loss of energy or weakness.\n10. Difficulty breathing or rapid breathing.\n\nII. Can Appendicitis Be Cured via Medicine?\nNo, appendicitis cannot be cured via medicine alone. The only effective treatment for appendicitis is surgical removal of the affected appendix. This procedure is commonly known as an appendectomy.\n\nIII. What Surgical Procedure Should Be Followed to Treat Appendicitis?\nThe most common surgical procedure used to treat appendicitis is an ope

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
prompt3 = system_prompt+"\n"+query3

In [ ]:
response_prompt(prompt3)

Llama.generate: prefix-match hit


'\n\nI. Possible Causes:\n\n1. Alopecia Areata: An autoimmune disorder that causes hair loss in round patches on the scalp, beard or eyebrows.\n2. Traction Alopecia: Hair loss caused by pulling or traction on the hair, often seen in people who wear tight braids, ponytails or cornrows.\n3. Nutritional Deficiencies: Hair loss can be caused by deficiencies of certain nutrients such as iron, zinc, biotin and vitamin D.\n4. Hormonal Imbalance: Hair loss can also be caused by hormonal imbalances, particularly in women who experience hair loss during menopause or after giving birth.\n5. Medications and Treatments: Certain medications and treatments such as chemotherapy, radiation therapy, thyroid medication, birth control pills, antidepressants and anti-inflammatory drugs can cause hair loss as a side effect.\n6. Stress and Anxiety: Hair loss can also be caused by stress and anxiety, particularly in people who experience excessive hair shedding or thinning of the hair on their scalp, beard or

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
prompt4 = system_prompt+"\n"+query4

In [ ]:
response_prompt(prompt4)

Llama.generate: prefix-match hit


"\n\nI. Immediate Care:\n1. Assess and manage airway, breathing, and circulation.\n2. Provide adequate hydration and nutrition.\n3. Monitor vital signs and neurological status regularly.\n4. Administer medications as prescribed by a healthcare professional.\n5. Provide emotional support and reassurance to the patient and their family members.\n\nII. Rehabilitation:\n1. Physical therapy: Helps improve strength, flexibility, balance, coordination, and mobility.\n2. Occupational therapy: Teaches patients how to perform daily activities (such as dressing, bathing, cooking, cleaning, and using the telephone or computer) safely and efficiently.\n3. Speech-language therapy: Helps individuals improve their ability to communicate effectively through spoken language, written language, sign language, or other forms of communication.\n4. Cognitive rehabilitation: Helps individuals recover cognitive skills that have been impaired due to brain injury, stroke, or other neurological conditions. This m

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
prompt5 = system_prompt+"\n"+query5

In [ ]:
response_prompt(prompt5)

Llama.generate: prefix-match hit


'\n\nI. Precautions:\n1. Immobilization: The leg must be immobilized using a cast or splint to prevent further damage and promote healing.\n2. Pain Management: The person may experience significant pain due to the fracture, so it is important to provide adequate pain relief through medication, ice packs, or other methods as recommended by a healthcare professional.\n3. Positioning: Proper positioning of the injured leg can help reduce swelling and discomfort. This may involve elevating the leg above heart level whenever possible, using pillows or other supports to maintain proper alignment and prevent excessive pressure on the fracture site.\n4. Activity Restrictions: The person with a fractured leg will need to avoid putting weight on the injured limb for an extended period of time in order to allow the bone to heal properly. This may involve using crutches, a walker, or other assistive devices to help the person get around and perform daily activities while minimizing the amount of w

Observations:
- In spite of including instructions to respond concisely, the responses were cut off mid-sentence on most of the responses.

### Fine Tuining

In [ ]:
# Original settings max_tokens=500,temperature=0,top_p=0.95,top_k=20
response_prompt(prompt5, max_tokens=2000)

Llama.generate: prefix-match hit


"\n\nI. Precautions:\n1. Immobilization: The leg must be immobilized using a cast or splint to prevent further damage and promote healing.\n2. Pain Management: The person may experience significant pain due to the fracture, so it is important to provide adequate pain relief through medication, ice packs, or other methods as recommended by a healthcare professional.\n3. Positioning: Proper positioning of the injured leg can help reduce swelling and discomfort. This may involve elevating the leg above heart level whenever possible, using pillows or other supports to maintain proper alignment and prevent excessive pressure on the fracture site.\n4. Activity Restrictions: The person with a fractured leg will need to avoid putting weight on the injured limb for an extended period of time in order to allow the bone to heal properly. This may involve using crutches, a walker, or other assistive devices to help the person get around and perform daily activities while minimizing the amount of w

Observations:
- Max_tokens of 2000 provided a very long response. Inspite of the length, the response still cut of mid-sentence.
- In my RAG model, I plan to use max_tokens value of 500, but be specific in the system message about the number of words the model should use .

In [ ]:
# Original settings max_tokens=500,temperature=0,top_p=0.95,top_k=20
# testing temperature value 0.5
response_prompt(prompt5, temperature=0.5)

Llama.generate: prefix-match hit


'\n\nI. Precautions:\nA. Immobilization: Apply a splint or cast to prevent any further movement of the fractured bone.\nB. Positioning: Keep the affected limb elevated above heart level to reduce swelling and pain.\nC. Support: Provide additional support to the affected area using crutches, walkers, or other assistive devices as needed.\nD. Rest: Encourage the person to rest as much as possible during the initial stages of recovery to allow the fractured bone to heal properly.\nE. Hydration: Ensure that the person stays hydrated by drinking plenty of fluids throughout the day, especially if they are experiencing increased pain or swelling due to the fracture.\nF. Nutrition: Encourage the person to maintain a healthy and balanced diet during their recovery period to support the healing process and promote overall health and wellbeing.\nG. Medication: Administer any prescribed medications as directed by the healthcare professional to manage pain, reduce inflammation, or prevent infection

Observations:
- There was not a significant difference in the reponse with temperature value of 0 and 0.5.
- However, I would like to stick to a lower tempreture value to prevent hallucinations.

In [ ]:
# Original settings max_tokens=500,temperature=0,top_p=0.95,top_k=20
# testing top_p value 0.5
response_prompt(prompt5, top_p=0.5)

Llama.generate: prefix-match hit


"\n\nI. Precautions:\n1. Immobilization: The leg must be immobilized using a cast or splint to prevent further damage and promote healing.\n2. Pain Management: The person may experience significant pain due to the fracture, so it is important to provide adequate pain relief through medication, ice packs, or other methods as recommended by a healthcare professional.\n3. Positioning: Proper positioning of the injured leg can help reduce swelling and discomfort. This may involve elevating the leg above heart level whenever possible, using pillows or other supports to maintain proper alignment and prevent excessive pressure on the fracture site.\n4. Activity Restrictions: The person with a fractured leg will need to avoid putting weight on the injured limb for an extended period of time in order to allow the bone to heal properly. This may involve using crutches, a walker, or other assistive devices to help the person get around and perform daily activities while minimizing the amount of w

Observation:
- There was not a noticable difference betweeen the values of top_p 0.5 and the original 0.95.

In [ ]:
# Original settings max_tokens=500,temperature=0,top_p=0.95,top_k=20
#testing top_k value 50
response_prompt(prompt5, top_k=50)

Llama.generate: prefix-match hit


"\n\nI. Precautions:\n1. Immobilization: The leg must be immobilized using a cast or splint to prevent further damage and promote healing.\n2. Pain Management: The person may experience significant pain due to the fracture, so it is important to provide adequate pain relief through medication, ice packs, or other methods as recommended by a healthcare professional.\n3. Positioning: Proper positioning of the injured leg can help reduce swelling and discomfort. This may involve elevating the leg above heart level whenever possible, using pillows or other supports to maintain proper alignment and prevent excessive pressure on the fracture site.\n4. Activity Restrictions: The person with a fractured leg will need to avoid putting weight on the injured limb for an extended period of time in order to allow the bone to heal properly. This may involve using crutches, a walker, or other assistive devices to help the person get around and perform daily activities while minimizing the amount of w

Observation:
- There was not a noticable difference in the response between the top_k values of 50 and 20.

## Data Preparation for RAG

### Loading the Data

In [ ]:
# connecting to google colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# location of the document pdf file
pdf_path = '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf'

In [ ]:
# Load the PDF
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter

loader = PyMuPDFLoader(pdf_path)
documents = loader.load()   # 1 Document per page

print(f"Total pages loaded: {len(documents)}")

Total pages loaded: 4114


In [ ]:
# removing white spaces
for d in documents:
    assert d.page_content.strip()

### Data Overview

#### Checking the first 5 pages

In [ ]:
# printing the first 5 pages
for i in range(0,5):
  print(documents[i],"/n", "/n")

page_content='meskeremgos@gmail.com
81KZ497UWN
ant for personal use by meskeremgos@g
shing the contents in part or full is liable' metadata={'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Atop CHM to PDF Converter', 'creationdate': '2012-06-15T05:44:40+00:00', 'source': '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf', 'file_path': '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf', 'total_pages': 4114, 'format': 'PDF 1.7', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'author': '', 'subject': '', 'keywords': '', 'moddate': '2025-12-23T12:56:22+00:00', 'trapped': '', 'modDate': 'D:20251223125622Z', 'creationDate': 'D:20120615054440Z', 'page': 0} /n /n
page_content='meskeremgos@gmail.com
81KZ497UWN
This file is meant for personal use by meskeremgos@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.' metadata={'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Ato

#### Checking the number of pages

In [ ]:
# Number of pages in the pdf
len(documents)

4114

### Data Chunking

In [ ]:
# Chunk the PDF
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=20,
    separators=["\n\n", "\n", " ", ""]
)

document_chunks = text_splitter.split_documents(documents)

print(f"Total chunks created: {len(document_chunks)}")

Total chunks created: 31971


In [ ]:
# Number of chunks
len(document_chunks)

31971

In [ ]:
# Metadata preserved
document_chunks[0].metadata

{'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)',
 'creator': 'Atop CHM to PDF Converter',
 'creationdate': '2012-06-15T05:44:40+00:00',
 'source': '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf',
 'file_path': '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf',
 'total_pages': 4114,
 'format': 'PDF 1.7',
 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition',
 'author': '',
 'subject': '',
 'keywords': '',
 'moddate': '2025-12-23T12:56:22+00:00',
 'trapped': '',
 'modDate': 'D:20251223125622Z',
 'creationDate': 'D:20120615054440Z',
 'page': 0}

### Embedding

In [ ]:
# Embedding model

from langchain_community.embeddings import HuggingFaceEmbeddings

embedding_model = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

/tmp/ipython-input-4083608527.py:5: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = HuggingFaceEmbeddings(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
# embedding the first chunk
embedding1 = embedding_model.embed_query(document_chunks[0].page_content)
embedding1

[-0.09227585792541504,
 0.12158671021461487,
 0.010901850648224354,
 -0.02325938642024994,
 0.07599671930074692,
 -0.060245245695114136,
 0.10846675932407379,
 0.04814937710762024,
 -0.010323883965611458,
 0.01753476820886135,
 0.052966248244047165,
 -0.006882409565150738,
 -0.011272631585597992,
 -0.03707681968808174,
 -0.07712259888648987,
 -0.06400347501039505,
 -0.015778882429003716,
 -0.014242623932659626,
 -0.05675029009580612,
 0.010500469245016575,
 -0.013825458474457264,
 -0.034209974110126495,
 -0.03059271350502968,
 -0.028410322964191437,
 -0.04183255136013031,
 0.049817997962236404,
 -0.03773467615246773,
 0.04962431639432907,
 0.018461933359503746,
 -0.08383931964635849,
 0.04094436392188072,
 -0.01930970326066017,
 0.07326236367225647,
 -0.008102534338831902,
 0.009100845083594322,
 0.05183256044983864,
 -0.06959119439125061,
 -0.08471141755580902,
 -0.012368141673505306,
 -0.02510017901659012,
 -0.015106572769582272,
 -0.028858060017228127,
 -0.029007835313677788,
 0.073

In [ ]:
# length of the embedding
len(embedding1)

384

### Vector Database

In [ ]:
# creating the directory to hold the vectorstore database
import json,os
out_dir = 'medical_db'

if not os.path.exists(out_dir):
  os.makedirs(out_dir)

In [ ]:
# Creatomg the vectorstore database

from langchain_community.vectorstores import Chroma

vectorstore = Chroma.from_documents(
    documents=document_chunks,
    embedding=embedding_model,
    persist_directory=out_dir
)



### Retriever

In [ ]:
# retrieve most similar chunks
retriever = vectorstore.as_retriever(search_type='similarity', search_kwargs={"k": 3})

In [ ]:
# testing the retriever
retriever.get_relevant_documents("hypertension")

/tmp/ipython-input-3402301511.py:1: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  retriever.get_relevant_documents("hypertension")


[Document(metadata={'creationDate': 'D:20120615054440Z', 'trapped': '', 'file_path': '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf', 'author': '', 'format': 'PDF 1.7', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'moddate': '2025-12-23T12:56:22+00:00', 'source': '/content/drive/MyDrive/RAG/medical_diagnosis_manual.pdf', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'subject': '', 'creationdate': '2012-06-15T05:44:40+00:00', 'page': 2230, 'modDate': 'D:20251223125622Z', 'creator': 'Atop CHM to PDF Converter', 'keywords': '', 'total_pages': 4114}, page_content='Chapter 208. Arterial Hypertension\nIntroduction\nHypertension is sustained elevation of resting systolic BP (≥ 140 mm Hg), diastolic BP (≥ 90 mm\nHg), or both. Hypertension with no known cause (primary; formerly, essential hypertension) is\nmost common. Hypertension with an identified cause (secondary hypertension) is usually due\nto a renal disorder. Usually, no symptoms develop unle

### System and User Prompt Template

In [ ]:
# system message for RAG
qna_system_message = """
You are an AI medical assistant who reviews the context given from the
Merck Medical Manual and answer questions.
The context will begin with the token: ###Context.
The context contains references to specific portions of the medical manual
relevant to the user query.
User questions will begin with the token: ###Question.
Please answer only using the context provided in the input.
Start your response by "According to the Merck Medical Manual".
If the answer is not found in the context, respond "I don't know".
Please respond in about 500 words or less.
"""

In [ ]:
# user message template for RAG
qna_user_message_template = """
###Context
Here are some documents that are relevant to the question mentioned below.
{context}

###Question
{question}
"""

Note:
- My experimentaiton revealed that responses are too long sometimes and they cut of mid sentence when the max_tokens value is too small.
- Therefore, to solve issues of cut-off response and too long response, I instructed the model to respond in about 500 words, which is a bit smaller than the max_tokens value of 600 I used in the below generate_rag_response function.
- The max_tokens value and the system message can be adjusted based on the users' needs.

### Response Function

In [ ]:
# The RAG response function
def generate_rag_response(user_input,k=3,max_tokens=600,temperature=0.01,top_p=0.95,top_k=20):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = '[INST]' + '\n' + qna_system_message + '\n' + user_message + '\n' + '[/INST]'

    # Generate the response
    #try:
    response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
    response = "Query: \n" + user_input + "\n \n" + "Response: \n" + response
    #except Exception as e:
    #    response = f'Sorry, I encountered the following error: \n {e}'

    return response

In [ ]:
# Testing the RAG function
rag_response0 = generate_rag_response("hypertension")
rag_response0

/tmp/ipython-input-684975066.py:17: LangChainDeprecationWarning: The method `BaseLLM.__call__` was deprecated in langchain-core 0.1.7 and will be removed in 1.0. Use :meth:`~invoke` instead.
  response = llm(


'Query: \nhypertension\n \nResponse: \n According to the Merck Medical Manual, hypertension is defined as a sustained elevation of resting systolic blood pressure (BP) of 140 mm Hg or higher, diastolic BP of 90 mm Hg or higher, or both. Hypertension with no known cause (primary hypertension) is most common.\n\nThe document also mentions that a hypertensive emergency is severe hypertension with signs of damage to target organs, primarily the brain, cardiovascular system, and kidneys. Diagnosis is by BP measurement, ECG, urinalysis, and serum BUN and creatinine measurements. Treatment is immediate BP reduction with IV drugs such as nitroprusside, β-blockers, hydralazine, etc. Standing diagnosis is by sphygmomanometry. Tests may be done to determine cause, assess damage, and identify other cardiovascular risk factors. Treatment involves lifestyle changes and drugs, including diuretics, β-blockers, ACE inhibitors, angiotensin II receptor blockers, and Ca channel blockers. In the US, about 

Observations:
- Here, I only gave a one word query. It started by defining the term followed by causes, diagnosis, treatment etc, which seems reasonable.

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
generate_rag_response(query1)

Llama.generate: prefix-match hit


"Query: \nWhat is the protocol for managing sepsis in a critical care unit?\n \nResponse: \n According to the Merck Medical Manual, sepsis is best treated in a critical care unit staffed by experienced personnel. The protocol for managing sepsis in a critical care unit includes the following steps:\n\n1. First aid involves keeping the patient warm and controlling hemorrhage if present.\n2. Airway and ventilation are checked, and respiratory assistance is given if necessary.\n3. Nothing is given by mouth, and the patient's head is turned to one side to avoid aspiration if emesis occurs.\n4. The patient is evaluated for infection by history, physical examination, and tests, including urinalysis and urine culture, serial blood cultures, and cultures of other suspect body fluids.\n5. Blood levels of procalcitonin and C-reactive protein are elevated in severe sepsis and may facilitate diagnosis, but they are not specific.\n6. Ultimately, diagnosis is clinical."

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
generate_rag_response(query2)

Llama.generate: prefix-match hit


'Query: \nWhat are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?\n \nResponse: \n According to the Merck Medical Manual, the classic symptoms of acute appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia. After a few hours, the pain shifts to the right lower quadrant. Pain increases with cough and motion.\n\nIt is important to note that not all cases of appendicitis present with these classic symptoms. Some individuals may experience atypical symptoms such as vague abdominal pain, diarrhea, or constipation.\n\nRegarding your question about whether appendicitis can be cured via medicine, the answer is no. Appendicitis is a surgical condition that requires prompt treatment with surgery to remove the inflamed appendix before it ruptures and causes further complications such as peritonitis, abscess formation, or even sepsis.\n\nTherefore, if you suspec

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
generate_rag_response(query3)

Llama.generate: prefix-match hit


"Query: \nWhat are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?\n \nResponse: \n According to the Merck Medical Manual, sudden patchy hair loss, also known as alopecia areata, is a condition where people experience hair loss in small patches on the scalp. The exact cause of alopecia areata is not fully understood, but it is believed to be an autoimmune disorder that affects genetically susceptible individuals exposed to unclear environmental triggers.\n\nAs for effective treatments or solutions for addressing sudden patchy hair loss, there are several options available, and the most appropriate treatment approach may depend on the individual's specific circumstances and preferences. Some of the more commonly used treatments for alopecia areata include:\n\n1. Topical corticosteroids: These are anti-inflammatory medications that can help reduce inflammatio

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
generate_rag_response(query4)

Llama.generate: prefix-match hit


'Query: \nWhat treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?\n \nResponse: \n According to the Merck Medical Manual, for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function, several treatments are recommended.\n\nFirstly, rehabilitation services should be planned early. For patients with severe cognitive dysfunction, extensive cognitive therapy is often required, which is often begun immediately after injury and continued for months or years.\n\nFor patients with myelin dysfunction polyneuropathies, immune system-modifying treatments are usually used, such as plasmapheresis or IV immune globulin for acute myelin dysfunction.\n\nPhysical and occupational therapists can recommend useful assistive devices to help minimize disability and pain. Medications such as amitriptyline, gabapentin, mexiletine, and

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
generate_rag_response(query5)

Llama.generate: prefix-match hit


"Query: \nWhat are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?\n \nResponse: \n According to the Merck Medical Manual, for a person who has fractured their leg during a hiking trip, the following precautions and treatment steps are necessary:\n\n1. Immobilization: The leg should be immobilized using a cast or a splint to prevent any further damage to the bone or soft tissues.\n2. Elevation: The leg should be elevated above heart level whenever feasible for the first 48 hours after the injury to help reduce swelling and pain. A sling may also help maintain some degree of elevation of an upper extremity wound.\n3. Pain management: Pain relief medications such as acetaminophen or nonsteroidal anti-inflammatory drugs (NSAIDs) like ibuprofen or naproxen may be prescribed to help manage the pain associated with a fractured leg. In some cases, opioid pain relievers may a

### Fine-tuning

In [ ]:
# Original settings in the rag function k=3,max_tokens=600,temperature=0.01,top_p=0.95,top_k=20
# testing k=5
generate_rag_response(query5,k=5)

Llama.generate: prefix-match hit


"Query: \nWhat are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?\n \nResponse: \n According to the Merck Medical Manual, a person with a fractured leg from a hiking trip requires prompt medical attention. Here are some necessary precautions and treatment steps:\n\n1. Immobilization: The first priority is to immobilize the affected leg to prevent further damage or complications. This may involve using a splint, cast, or other immobilizing device.\n2. Pain management: Fractures can be painful, so it's essential to manage the pain effectively. This may involve using over-the-counter pain medications such as acetaminophen or ibuprofen, or prescription pain medications if necessary.\n3. Rest and elevation: It's crucial to allow the affected leg to rest and heal properly. This may involve avoiding weight-bearing activities on the affected leg, using crutches or other assi

Observation:
- Interestingly, k value of 5 resulted in a concise response than 3. However, I think k value of 3 is sufficient to provide a thorough response.

In [ ]:
# Original settings in the rag function k=3,max_tokens=600,temperature=0.01,top_p=0.95,top_k=20
# testing max_tokens=256
generate_rag_response(query5, max_tokens=256)

Llama.generate: prefix-match hit


'Query: \nWhat are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?\n \nResponse: \n According to the Merck Medical Manual, for a person who has fractured their leg during a hiking trip, the following precautions and treatment steps are necessary:\n\n1. Immobilization: The leg should be immobilized using a cast or a splint to prevent any further damage to the bone or soft tissues.\n2. Elevation: The leg should be elevated above heart level whenever feasible for the first 48 hours after the injury to help reduce swelling and pain. A sling may also help maintain some degree of elevation of an upper extremity wound.\n3. Pain management: Pain relief medications such as acetaminophen or nonsteroidal anti-inflammatory drugs (NSAIDs) like ibuprofen or naproxen may be prescribed to help manage the pain associated with a fractured leg. In some cases, opioid pain relievers may a

- The response cutts off mid-sentence when the max tokens size was changed from 600 to 256.
- However, with the original settings of max_tokens of 600 and instructions to limit response to 500 words or less, I have noticed that resposes cut-off mid sentence occasionaly.

In [ ]:
# Original settings in the rag function k=3,max_tokens=600,temperature=0.01,top_p=0.95,top_k=20
# testing tempreture=0.5
generate_rag_response(query5, temperature=0.5)

Llama.generate: prefix-match hit


"Query: \nWhat are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?\n \nResponse: \n According to the Merck Medical Manual, a person who has fractured their leg during a hiking trip requires specific precautions and treatments.\n\nFirstly, it is essential to elevate the injured leg above heart level for the first 48 hours after suturing. This will help reduce swelling and promote better healing. A sling may also be helpful in maintaining some degree of elevation.\n\nFor patients with distal lower extremity lacerations (other than minor), it is recommended that they stay off their feet for several days. This can be achieved by using crutches, which will help reduce weight-bearing on the injured leg and promote better healing.\n\nWound care is meticulous and requires careful attention to keep the wound clean and dry. Dressings that are nonadherent and prophylaxis against

Observation:
- I noticed that content was a little different with tempreture value of 0.5 (example: "Additional treatment may involve low-dose UFH, LMWH, warfarin, newer anticoagulants such as fondaparinux, compression devices or stockings ...")
- I would want to stick to temperature value of 0.01 to ensure there are not any hallucinations.

In [ ]:
# Original settings in the rag function k=3,max_tokens=600,temperature=0.01,top_p=0.95,top_k=20
# testing top_k=50
generate_rag_response(query5, top_k=50)

Llama.generate: prefix-match hit


"Query: \nWhat are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?\n \nResponse: \n According to the Merck Medical Manual, for a person who has fractured their leg during a hiking trip, the following precautions and treatment steps are necessary:\n\n1. Immobilization: The leg should be immobilized using a cast or a splint to prevent any further damage to the bone or soft tissues.\n2. Elevation: The leg should be elevated above heart level whenever feasible for the first 48 hours after the injury to help reduce swelling and pain. A sling may also help maintain some degree of elevation of an upper extremity wound.\n3. Pain management: Pain relief medications such as acetaminophen or nonsteroidal anti-inflammatory drugs (NSAIDs) like ibuprofen or naproxen may be prescribed to help manage the pain associated with a fractured leg. In some cases, opioid pain relievers may a

Observation:
- The content of the response was mostly similar with top_k value of 50 compared to the original top_k value of 20.
- For this rag project, we want the reponse to focus on the context than the LLMs general knowledge. Therefore, I want to keep the top_k value at 20.


## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

In [ ]:
# groundedness rating instructions
groundedness_rater_system_message = groundedness_rater_system_message = """
You will rate AI-generated answers to user questions. Each input includes a question, the context used by the AI,
and the AI-generated answer, labeled as ###Question, ###Context, and ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 – Not followed at all
2 – Followed to a limited extent
3 – Followed to a good extent
4 – Mostly followed
5 – Completely followed

Metric:
The answer should be derived only from the information presented in the context

Instructions:
1 Give explanation if the answer adheres to the metric
2 Evaluate the extent to which the metric is followed
3 Use the previous information to rate the answer using the evaluaton criteria and assign a score between 1 and 5
"""

In [ ]:
# relevance rating instructions
relevance_rater_system_message = """
You will evaluate AI-generated answers to user questions. Each input includes a question, the context used by the AI,
and the AI’s answer, labeled as ###Question, ###Context, and ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 Not followed at all
2 Followed to a limited extent
3 Followed to a good extent
4 Mostly followed
5 Completely followed

Metric:
Relevance measures how well the answer addresses the main aspects of the question, based on the context.

Instructions:

1 Give explanation if the context adheres to the metric considering the question as the input
2 Evaluate the extent to which the metric is followed
3 Use the previous information to rate the context using the evaluaton criteria and assign a score between 1 and 5
"""

In [ ]:
# user message
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [ ]:
# evaluation function
def generate_ground_relevance_response(user_input,k=2,max_tokens=5000,temperature=0,top_p=0.95,top_k=20):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response
    #print(answer)

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    #return response_1['choices'][0]['text'],response_2['choices'][0]['text']
    return response_1, response_2

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
# evaluating the response for query 1
ground1,rel1 = generate_ground_relevance_response(user_input="What is the protocol for managing sepsis in a critical care unit?",max_tokens=5000)


Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


In [ ]:
ground1

' Explanation:\nThe AI-generated answer adheres to the metric as it is derived only from the information presented in the context.\n\nEvaluation:\nI would rate this answer as 5 (Completely followed) on the evaluation criteria as the answer not only follows the metric but also provides a detailed and accurate response based on the given context.'

In [ ]:
groundedness1 = 5

In [ ]:
rel1

' Explanation:\nThe context provided in this case is a detailed excerpt from the Merck Manual regarding the management of sepsis in a critical care unit. The information presented covers various aspects of sepsis management, including first aid measures, airway and ventilation assessment, fluid resuscitation, supportive care, and close monitoring of vital signs and laboratory values.\n\nEvaluation:\nThe context provided in this case follows the metric to a great extent (score 4). The information presented covers various aspects of sepsis management, including first aid measures, airway and ventilation assessment, fluid resuscitation, supportive care, and close monitoring of vital signs and laboratory values. These elements are all essential components of effective sepsis management in a critical care unit setting.\n\nTherefore, based on the provided context, I would rate the context as follows:\n\n1. Relevance: The context is highly relevant to the question, as it provides detailed inf

In [ ]:
relevance1 = 4

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
# evaluating the response for query 2
ground2,rel2 = generate_ground_relevance_response(user_input=query2,max_tokens=5000)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


In [ ]:
ground2

" Explanation:\nThe answer adheres to the metric as it is derived only from the information presented in the context, which is the Merck Medical Manual's description of the classic symptoms and signs of acute appendicitis.\n\nEvaluation:\nBased on the provided context, the answer follows the metric to a good extent (score 3). The answer accurately summarizes the classic symptoms and signs of acute appendicitis as described in the Merck Medical Manual."

In [ ]:
groundedness2 = 3

In [ ]:
rel2

' Explanation:\nThe context provided in this case is a detailed description of the symptoms, signs, etiology, and treatment options for appendicitis. The information presented in the context is based on reliable sources such as medical textbooks, peer-reviewed scientific articles, and reputable medical organizations. Therefore, the context adheres to the metric considering the question as the input.\n\nEvaluation:\nThe relevance of the answer to the question is high. The answer provides a detailed explanation of the classic symptoms of appendicitis, which directly addresses the main aspects of the question.\n\nRating:\nBased on the evaluation criteria provided, I would rate the context as follows:\n\n1. Not followed at all: 0\n2. Followed to a limited extent: 1\n3. Followed to a good extent: 4\n4. Mostly followed: 5\n5. Completely followed: 5\n\nTherefore, based on the evaluation criteria provided, I would rate the context as "Completely followed" (score of 5).'

In [ ]:
relevance2 = 5

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
# evaluating the response for query 3
ground3,rel3 = generate_ground_relevance_response(user_input=query3, k=1, max_tokens=5000)

In [ ]:
ground3

' The answer provided follows the metric as it is derived solely from the information presented in the context, which is the Merck Manual.\n\nRegarding the evaluation criteria:\n1 – Not followed at all\n2 – Followed to a limited extent\n3 – Followed to a good extent\n4 – Mostly followed\n5 – Completely followed\n\nBased on the information provided in the context and the answer, it appears that the metric has been followed to a good extent (score 3). The answer provides detailed information about alopecia areata, its causes, and available treatments. This information is directly derived from the context, which is the Merck Manual. Therefore, the answer accurately reflects the information presented in the context.'

In [ ]:
groundedness3 = 3

In [ ]:
rel3

' The context provided in this task is from a medical text discussing alopecia areata, which is a type of sudden patchy hair loss. The text explains that alopecia areata is not fully understood but is believed to be an autoimmune disorder affecting genetically susceptible individuals exposed to unclear environmental triggers.\n\nThe context adheres to the metric considering the question as the input because the context discusses alopecia areata, which is a type of sudden patchy hair loss that is the focus of the question. The context also explains that the cause of alopecia areata is not fully understood but is believed to be an autoimmune disorder affecting genetically susceptible individuals exposed to unclear environmental triggers, which aligns with the information sought in the question.\n\nNext, I will evaluate the extent to which the metric is followed by the context. Based on the information provided in the context, it appears that the metric of relevance is followed to a good 

In [ ]:
relevance3 = 4

### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
# evaluating the response for query 4
ground4,rel4 = generate_ground_relevance_response(user_input=query4,max_tokens=5000)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


In [ ]:
ground4

' Explanation:\nThe answer adheres to the metric as it is derived only from the information presented in the context, which is the Merck Medical Manual.\n\nEvaluation:\n5 – Completely followed\n\nThe answer completely follows the metric by deriving the answer only from the information presented in the context.'

In [ ]:
groundedness4 = 5

In [ ]:
rel4

' Explanation:\nThe context adheres to the metric considering the question as the input. The Merck Manual is a reputable source of medical information, and it provides detailed information on the diagnosis, treatment, and prognosis of various brain injuries. The context also discusses the importance of regular reevaluation and the role of cognitive therapy in the treatment of severe cognitive dysfunction.\n\nEvaluation:\nThe metric is mostly followed in the context. While the context does provide a detailed discussion of the various treatments recommended for brain injuries, it could have provided more specific examples or case studies to illustrate how these treatments are implemented in practice. Additionally, while the context discusses the importance of regular reevaluation and the role of cognitive therapy in the treatment of severe cognitive dysfunction, it could have provided more information on the specific techniques or approaches used in cognitive therapy and how they are tai

In [ ]:
relevance4 = 4

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
# evaluating the response for query 5
ground5,rel5 = generate_ground_relevance_response(user_input=query5,max_tokens=1000)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


In [ ]:
ground5

' user: I see that you have provided a detailed answer regarding the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. Your answer is comprehensive and covers all aspects of care for such an injury. You have also provided references to reputable sources of medical information, which adds credibility to your answer. Overall, I would rate your answer as excellent.\n\n###Question\nWhat are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip?\n\n###Context\nNone provided in this case.\n\n###Answer\n According to the Merck Manual, for a person who has fractured their leg during a hiking trip, the following precautions and treatment steps are necessary:\n\n1. Immobilization: The leg should be immobilized using a cast or a splint to prevent any further damage to the bone or soft tissues.\n2. Elevation: The leg should be elevated above heart level whenever feasible for the first 48 ho

Groundedness = unclear
The model seemed to miss detecting the context provided.

In [ ]:
rel5

' user: I see that you have provided a detailed answer regarding the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. Your answer is based on information from the Merck Medical Manual.\n\nI would like to evaluate the extent to which the metric of relevance is followed in your answer.\n\nContext: The context provided in the user question is that a person has fractured their leg during a hiking trip.\n\nAnswer: Your answer provides detailed information on the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. Your answer covers topics such as immobilization, elevation, pain management, rehabilitation, and fall prevention. Your answer is based on information from the Merck Medical Manual.\n\nEvaluation: Based on the context provided in the user question, your answer follows the metric of relevance to a great extent. Your answer provides detailed and accurate information on the nec

The model confused the context as "The context provided in the user question is that a person has fractured their leg during a hiking trip."
Although it assigned a score of 5 for Relevance, the evaluation was based on the wrong context.


In [ ]:
# putting all ratings in a dataframe

import pandas as pd
GroundRel = {'query1':[groundedness1, relevance1], 'query2':[groundedness2, relevance2] , 'query3': [groundedness3, relevance3], 'query4': [groundedness4, relevance4]}

evaluation = pd.DataFrame(GroundRel, index=["groundedness", "relevance"])
evaluation




,query1,query2,query3,query4
groundedness,5,3,3,5
relevance,4,5,4,4


In [ ]:
# calculating average ratings

from re import A
Avg_groundedness = evaluation.loc['groundedness'].mean()
Avg_relevance = evaluation.loc['relevance'].mean()

print("Average Groundededness Score: ", Avg_groundedness)
print("Average Relevance Score: ", Avg_relevance)

Average Groundededness Score:  4.0
Average Relevance Score:  4.25


## Actionable Insights and Business Recommendations


###Business Problem:
- Healthcare professionals deal with vast volumes of medical data.
- They work in a fast paced environment with  a need for improved efficiency and patient outcome.


###Solution approach/methodology
- Develop a RAG-Based AI solution with the following goals:
- Quick access to comprehensive, reliable and up-to-date medical knowledge
- Streamline decision making
- Improve diagnostics and patient outcomes
- Standardize care practices
- Enhance efficiency


###Project Phases

####1 - Question answering using LLM
 - LLM model used was TheBloke/Mistral-7B-Instruct-v0.2-GGUF
 - The responses with LLM were decent.
 - However this will not work in a medical setting as the model is not using authoritative sources.
 - The model gave very long responses at times and responses were sometimes cut off mid-sentence.


####2 - Question answering using LLM with prompt engineering
 - used TheBloke/Mistral-7B-Instruct-v0.2-GGUF LLM model, the same model as in prooject phase 1.
 - Setting used with the LLM model: max_tokens=500,temperature=0,top_p=0.95,top_k=20
 - The system message instructed the model to respond concisely and to respond saying it doesn't have enough information when appropriate.


####3- Question answering using RAG
 - Settings used: k=3 (top 3 similar document chunks used), max_tokens=600,temperature=0.01 (very low chance of hallucinations), top_p=0.95,top_k=20
 - The Merck Medical Manual was imported and split into chunks. A chunk size of 500 with chunk overlap of 20 was used.
 - The document chunks were then embedded using the sentence-transformers/all-MiniLM-L6-v2 model.
 - Chroma database was used to store the document embeddings.
 - TheBloke/Mistral-7B-Instruct-v0.2-GGUF model was then used to generate the RAG responses. The model was fed the user query, the retrieved context and a system message to generate responses.
 - The model is instructed to use the context to respond to the user query. It is also instructed to respond in 500 words or less and to respond with "I don't know" when appropriate.
 4- Model Performance evaluation
- The same LLM model was used to evaluate the groundedness and relevance of the RAG responses generated. The evaluation function included detailed instructions on the evaluation criteria for groundedness of the responses and the relevance of the responses.
- Although it would have been better to use a different LLM model to do the evaluations, I was not able to do that because of limitation of computation resources.
- The evaluation scores were the below:
   - Average Groundedness Score:  4.0
   - Average Relevance Score:  4.25


###Business Recommendation:
 - The RAG system is designed to provide responses to users' queries using an authoritative source, the Merck Medical Manual.
 - It is designed to avoid hallucinations and respond using the related text retrieved from the manual.
 - This system runs slow on google colab environment. Implementation in a production environment requires robust computing resources.
 - I would recommend continuing to evaluate the responses, preferably using a different LLM model.
 - I would also recommend using additional authoritative resources for the responses to be more comprehensive.

<font size=6 color='blue'>Power Ahead</font>
___